README
App::DH

App::DH is a basic skeleton of a command line interface for the
excellent "DBIx::Class::DeploymentHandler", to make executing database
deployment stages easier.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by The App::DH Authors,
Contributors, and Sponsors.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for App-DH

{{$NEXT}}
 [Bugfix]
 - Fix tests failing with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.004000, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: ↓1

0.004000 2016-10-08T05:23:17Z d3631d1
 [Commands]
 - New commands "database_version" and "schema_version" for inspecting relevant version fields.
 - Thanks to Daniel Böhmer for PR #2

 [Dependencies::Stats]
 - Dependencies changed since 0.003000, see misc/*.deps* for details
 - configure: (suggests: +1)
 - develop: +4 ↑1 (suggests: ↑1)

0.003000 2015-06-27T04:31:29Z 3f38f2a
 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - develop: +1 -1 (suggests: ↑2)
 - runtime: +1
 - test: +6 (recommends: +1)

 [Features]
 - Can now specify a target version with --target
 - Thanks to Yanick Champoux #1

0.002000 2015-03-08T09:13:15Z 2350202
 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +52 -2 (recommends: +1 -1, suggests: +1 ↑1)
 - runtime: +2 -1 (recommends: +1)
 - test: +3 ↓1 -2 (recommends: +3)

 [Internals]
 - A few dependencies that were no longer needed have been shaved from runtime.

 [Meta]
 - Move repository to kentnl org
 - Updated git URLs

 [Packaging]
 - Switched to EUMM: https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f
 - Dependencies on toolkit softened.

0.001002 2013-04-07T22:12:52Z
 [Bugfix]
 - Depend on MooseX::Getopt 0.48 because we wrap one of the methods added at that version.

 [Dependencies::Changed / build requires]
 - Module::Build 0.4003 → 0.4004

 [Dependencies::Changed / configure requires]
 - Module::Build 0.4003 → 0.4004

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.7.0 → v1.7.2

 [Dependencies::Changed / runtime requires]
 - MooseX::Getopt 0 → 0.48

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: (suggests: ↑1)
 - runtime: ↑1

 [Packaging]
 - Utilise Module::Build 0.4004 test_requires

0.001001 2013-02-15T16:09:39Z
 - default databases are now chosen by introspecting your --connection_name. But you can still override the default on
   the commandline.

0.001000 2013-02-15T03:56:31Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Deploy your DBIx::Class Schema to DDL/Database via DBIx::Class::DeploymentHandler",
  "AUTHOR" => "kentnl - Kent Fredric (cpan:KENTNL) <kentnl\@cpan.org>, mst - Matt S. Trout (cpan:MSTROUT) <mst\@shadowcat.co.uk>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-DH",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "App::DH",
  "PREREQ_PM" => {
    "Carp" => 0,
    "DBIx::Class::DeploymentHandler" => 0,
    "Module::Runtime" => 0,
    "Moose" => 0,
    "MooseX::Getopt" => "0.48",
    "PerlX::Maybe" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "DBI" => 0,
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.89",
    "Test::Requires" => 0,
    "Test::TempDir::Tiny" => 0,
    "base" => 0
  },
  "VERSION" => "0.004900",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "DBI" => 0,
  "DBIx::Class::Core" => 0,
  "DBIx::Class::DeploymentHandler" => 0,
  "DBIx::Class::Schema" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Runtime" => 0,
  "Moose" => 0,
  "MooseX::Getopt" => "0.48",
  "PerlX::Maybe" => 0,
  "Test::More" => "0.89",
  "Test::Requires" => 0,
  "Test::TempDir::Tiny" => 0,
  "base" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/App/DH.pm
use 5.006;    # our
use strict;
use warnings;

package App::DH;

our $VERSION = '0.004900';

# ABSTRACT: Deploy your DBIx::Class Schema to DDL/Database via DBIx::Class::DeploymentHandler

# AUTHORITY

use Carp qw( croak );
use DBIx::Class::DeploymentHandler;
use Moose qw( with has around );
use MooseX::Getopt 0.48 ();
use PerlX::Maybe qw( maybe );

with 'MooseX::Getopt';

=param --connection_name

    -c/--connection_name

Specify the connection details to use for deployment.
Can be a name of a configuration in a C<DBIx::Class::Schema::Config> configuration if the L</--schema> uses it.

    --connection_name 'dbi:SQLite:/path/to/db'

    -cdevelopment

=cut

has connection_name => (
  traits        => ['Getopt'],
  is            => ro =>,
  isa           => Str =>,
  required      => 1,
  cmd_aliases   => c =>,
  default       => sub { 'development' },
  documentation => 'either a valid DBI DSN or an alias configured by DBIx::Class::Schema::Config',
);

=param --force

Overwrite existing DDL files of the same version.

    -f/--force

=cut

has force => (
  traits        => ['Getopt'],
  is            => ro =>,
  isa           => Bool =>,
  default       => sub { 0 },
  cmd_aliases   => f =>,
  documentation => 'forcefully replace existing DDLs. [DANGER]',
);

=param --schema

    -s/--schema

The class name of the schema to load for DDL/Deployment

    -sMyProject::Schema
    --schema MyProject::Schema

=cut

has schema => (
  traits        => ['Getopt'],
  is            => ro =>,
  isa           => Str =>,
  required      => 1,
  cmd_aliases   => s =>,
  documentation => 'the class name of the schema to generate DDLs/deploy for',
);

=param --include

    -I/--include

Add a given library path to @INC prior to loading C<schema>

    -I../lib
    --include ../lib

May be specified multiple times.

=cut

has include => (
  traits        => ['Getopt'],
  is            => ro =>,
  isa           => ArrayRef =>,
  default       => sub { [] },
  cmd_aliases   => I =>,
  documentation => 'paths to load into INC',
);

=param --script_dir

    -o/--script_dir

Specify where to write the per-backend DDL's.

Default is ./share/ddl

    -o/tmp/ddl
    --script_dir /tmp/ddl

=cut

has script_dir => (
  traits        => ['Getopt'],
  is            => ro =>,
  isa           => Str =>,
  default       => sub { 'share/ddl' },
  cmd_aliases   => o =>,
  documentation => 'output path',
);

=param --database

    -d/--database

Specify the C<SQL::Translator::Producer::*> backend to use for generating DDLs.

    -dSQLite
    --database PostgreSQL

Can be specified multiple times.

Default is introspected from looking at whatever L</--connection_name> connects to.

=cut

has database => (
  traits        => ['Getopt'],
  is            => 'ro',
  lazy          => 1,
  builder       => '_build_database',
  isa           => ArrayRef =>,
  cmd_aliases   => d =>,
  documentation => 'SQL::Translator::Producer::* database backends to generate DDLs for',
);

=param --target

    --target

Specify which version to install/upgrade to.

If not specified, defaults to the latest version.

=cut

has target => (
  traits        => ['Getopt'],
  isa           => 'Str',
  is            => 'ro',
  predicate     => 'has_target',
  cmd_aliases   => 'v',
  documentation => 'version to install/upgrade to',
);

has _dh     => ( is => 'ro', lazy => 1, builder => '_build__dh' );
has _schema => ( is => 'ro', lazy => 1, builder => '_build__schema' );

sub _build__schema {
  my ($self) = @_;
  require lib;
  lib->import($_) for @{ $self->include };
  require Module::Runtime;
  my $class = Module::Runtime::use_module( $self->schema );
  return $class->connect( $self->connection_name );
}

sub _build__dh {
  my ($self) = @_;
  return DBIx::Class::DeploymentHandler->new(
    {
      schema           => $self->_schema,
      force_overwrite  => $self->force,
      script_directory => $self->script_dir,
      databases        => $self->database,
      maybe to_version => $self->target,
    },
  );
}

sub _build_database {
  my ($self) = @_;
  my $type = $self->_schema->storage->sqlt_type;

  # Note: This seemingly needless stringification
  # exists to solve an incredibly complex problem on bleadperl
  # with COW, and for some reason, the string sqlt_type
  # returns as the flag 'IsCOW',
  # which for some reason causes a warning when the invoking
  # perl interpreter terminates.
  #
  # If you can solve the bug in bleadperl, I'll
  # gladly remove the forced stringification of the COW string.
  #
  # -- kentnl @ Feb 16/2013
  # -- perl (v5.17.9 (v5.17.8-156-g012528a))
  return ["$type"];
}

=cmd write_ddl

Only generate ddls for deploy/upgrade

    dh.pl [...params] write_ddl

=cut

sub cmd_write_ddl {
  my ($self) = @_;
  $self->_dh->prepare_install;
  my $v = $self->_dh->schema_version;
  if ( $v > 1 ) {
    $self->_dh->prepare_upgrade(
      {
        from_version => $v - 1,
        to_version   => $v,
      },
    );
  }
  return;
}

=cmd install

Install to connection L</--connection_name>

    dh.pl [...params] install

=cut

sub cmd_install {
  my $self = shift;
  $self->_dh->install;
  return;
}

=cmd upgrade

Upgrade connection L</--connection_name>

    dh.pl [...params] upgrade

=cut

sub cmd_upgrade {
  my $self = shift;
  $self->_dh->upgrade;
  return;
}

=cmd database_version

Report database_version of L</--connection>

=cut

sub cmd_database_version {
    my $self = shift;

  ## no critic (RequireCheckedSyscalls)
  print $self->_dh->database_version . "\n";

  return;
}

=cmd schema_version

Report schema_version of L</--schema>

    dh.pl [...params] schema_version

=cut

sub cmd_schema_version {
  my $self = shift;

  ## no critic (RequireCheckedSyscalls)
  print $self->_dh->schema_version . "\n";

  return;
}

my (%cmds) = (
  write_ddl        => \&cmd_write_ddl,
  install          => \&cmd_install,
  upgrade          => \&cmd_upgrade,
  database_version => \&cmd_database_version,
  schema_version   => \&cmd_schema_version,
);
my (%cmd_desc) = (
  write_ddl        => 'only write ddl files',
  install          => 'install to the specified database connection',
  upgrade          => 'upgrade the specified database connection',
  database_version => 'report the version of the specified database connection',
  schema_version   => 'report the version of the schema class',
);
my $list_cmds = join q[ ], sort keys %cmds;
my $list_cmds_opt = '(' . ( join q{|}, sort keys %cmds ) . ')';
my $list_cmds_usage =
  ( join qq{\n}, q{}, qq{\tcommands:}, q{}, ( map { ( sprintf qq{\t%-30s%s}, $_, $cmd_desc{$_} ) } sort keys %cmds ), q{} );

=begin Pod::Coverage

    cmd_write_ddl
    cmd_install
    cmd_upgrade
    cmd_database_version
    cmd_schema_version
    run

=end Pod::Coverage

=cut

around print_usage_text => sub {
  my ( undef, undef, $usage ) = @_;
  my ($text) = $usage->text();
  $text =~ s{
        ( long\s+options[.]+[]] )
    } {
        $1 . ' ' . $list_cmds_opt
    }msex;
  $text .= qq{\n} . $text . $list_cmds_usage . qq{\n};
  print $text or croak q[Cannot write to STDOUT];
  exit 0;
};

sub run {
  my ($self) = @_;
  my ( $cmd, @what ) = @{ $self->extra_argv };
  croak "Must supply a command\nCommands: $list_cmds\nFailed" unless $cmd;
  croak "Extra argv detected - command only please\nFailed" if @what;
  croak "No such command ${cmd}\nCommands: $list_cmds\nFailed"
    unless exists $cmds{$cmd};
  my $code = $cmds{$cmd};
  return $self->$code();
}

__PACKAGE__->meta->make_immutable;
no Moose;

1;

=head1 SYNOPSIS

Basic usage:

    #!/usr/bin/env perl
    #
    # dh.pl

    use App::DH;
    App::DH->new_with_options->run;

--

	usage: dh.pl [-?cdfhIos] [long options...] (install|upgrade|write_ddl)
		-h -? --usage --help     Prints this usage information.
		-c --connection_name     either a valid DBI DSN or an alias
		                         configured by DBIx::Class::Schema::Config
		-f --force               forcefully replace existing DDLs. [DANGER]
		-s --schema              the class name of the schema to generate
		                         DDLs/deploy for
		-I --include             paths to load into @INC
		-o --script_dir          output path
		-d --database            database backends to generate DDLs for. See
		                         SQL::Translator::Producer::* for valid values

		commands:

		install                       install to the specified database connection
		upgrade                       upgrade the specified database connection
		write_ddl                     only write ddl files


If you don't like any of the defaults, you can subclass to override

    use App::DH;
    {
        package MyApp;
        use  Moose;
        extends 'App::DH';

        has '+connection_name' => ( default => sub { 'production' } );
        has '+schema'          => ( default => sub { 'MyApp::Schema' } );
        __PACKAGE__->meta->make_immutable;
    }
    MyApp->new_with_options->run;

=head1 DESCRIPTION

App::DH is a basic skeleton of a command line interface for the excellent
L<< C<DBIx::Class::DeploymentHandler>|DBIx::Class::DeploymentHandler >>, to make executing database deployment stages easier.

=head1 CREDITS

This module is mostly code by mst, sponsored by L<nordaaker.com|http://nordaaker.com>, and I've only tidied it up and made it
more CPAN Friendly.

=head1 SPONSORS

The authoring of the initial incarnation of this code is kindly sponsored by L<nordaaker.com|http://nordaaker.com>.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw(ddls DDL backend DDL's DDLs mst nordaaker CPAN kentnl));
for my $var (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $var ) );
}
$bundle->add_or_append_policy_field(
  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.032.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4211"
  },
  "configure_requires" => {
    "Module::Build" => "0.4211",
    "perl" => "5.006"
  },
  "dist_abstract" => "Deploy your DBIx::Class Schema to DDL/Database via DBIx::Class::DeploymentHandler",
  "dist_author" => [
    "kentnl - Kent Fredric (cpan:KENTNL) <kentnl\@cpan.org>",
    "mst - Matt S. Trout (cpan:MSTROUT) <mst\@shadowcat.co.uk>"
  ],
  "dist_name" => "App-DH",
  "dist_version" => "0.001003",
  "license" => "perl",
  "module_name" => "App::DH",
  "recommends" => {
    "DBIx::Class::Schema::Config" => 0,
    "Moose" => "2.000"
  },
  "recursive_test_files" => 1,
  "requires" => {
    "DBIx::Class::DeploymentHandler" => 0,
    "Module::Runtime" => 0,
    "Moose" => 0,
    "MooseX::AttributeShortcuts" => 0,
    "MooseX::Getopt" => "0.48",
    "lib" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.301001_098",
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4211",
  "Test::More" => "1.301001_098",
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
